File: joptimize/test/src/examples/narrow/NarrowReturn.scala
package joptimize.examples.narrow

object NarrowReturn {
  def main(x: Int, y: Int): Int = {
    call(wrap(Bar), x) + call(wrap(Qux), y)
  }

  def call(f: Foo, n: Int): Int = f.inc(n)

  def wrap(f: Foo): Foo = f
}

File: joptimize/test/src/examples/narrow/NarrowReturn.java
package joptimize.examples.narrow;

class NarrowReturn {
    static int main(int x, int y){
    return call(wrap(new Bar()), x) + call(wrap(new Qux()), y);
  }

    static int call(Foo f, int n){ return f.inc(n); }

  static Foo wrap(Foo f){ return f; }

    interface Foo{
        int inc(int n);
    }
    static class Bar implements Foo{
        public int inc(int n){ return n + 1; }
    }
    static class Qux implements Foo{
        public int inc(int n){ return n + 2; }
    }
}
